### PR TITLE
Switch cluster references in placement from strings to objects

### LINKED ISSUE #688: Placement references to clusters violates API convention
TLDR: Cluster names for the purposes of placement are specified unqualified, but according to kube api conventions that means a given cluster should reside in the same namespace as the federated resource that refers to it.

### Background Info

In the history of federation-v2, the `Cluster` resource in the cluster-registry and `FederatedCluster` resource in federation-v2 were originally cluster-scoped. This was reflective of the scoping of the `Cluster` resource in the original federation effort. Another historical factor is the identification of clusters in federation APIs by name only, which was entirely sufficient due to their cluster-scoped nature. Finally, originally federation-v2 was conceived of as running exclusively in a cluster-scoped mode where a single controller serviced all the namespaces in a cluster. 

Later in time, three things changed:

- The `Cluster` resource was changed to be namespace-scoped
- Correspondingly, the `FederatedCluster` resource was changed to be namespace-scoped
- We introduced the ability to deploy the federation controllers scoped to a single namespace

These changes introduced a wrinkle from an API perspective: the semantics of identifying clusters in federated APIs depend on how federation is deployed:

- When federation is deployed scoped to a namespace, `spec.placement.clusterNames` is a list of the names of `FederatedClusters` in the containing namespace
- When federation is deployed scoped to the entire cluster, `spec.placement.clusterNames` is a list of the names of `FederatedClusters` in a namespace whose value is a configuration item of the federation controller.

According to the [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references) guidance on object references, object references made using just the name should correspond to objects within the containing namespace. We adhere to this guidance when federation is deployed scoped to a namespace; we don't adhere when federation is deployed in the cluster scope.

A simple thing that we can do to adhere to this guidance for ns-scoped types is for name-only references always to refer to things within the containing namespace. While this would require changes to the controller, it would not affect the existing API surface for ns-scoped resources.

### Example

Let's consider the example of a `FederatedDeployment` in the `example` namespace:

```yaml
apiVersion: types.federation.k8s.io/v1alpha1
kind: FederatedDeployment
metadata:
  name: hello
  namespace: example
spec:
  placement:
    clusterNames:
    - cluster-x
    - cluster-y
  # rest omitted
```

The `FederatedClusters` addressed by `spec.placement.clusterNames` are ambiguous; it might be:

- `cluster-x` and `cluster-y` in the `example` namespace
- `cluster-x` and `cluster-y` in the configured federation namespace of the controller

### PR DESCRIPTION
**Pull request by marun**
This ensures that the addition of fields to cluster references will be backwards compatible.

Fixes #688

### CODE REVIEW

**Comment by shashidharatd** [comment]
LGTM

**Comment by pmorie** [comment]
I'm lgtm on this.

**Comment by shashidharatd** [comment]
@marun, this needs rebase.

**Comment by marun** [comment]
Rebased

File: docs/userguide.md [hunk 1]
```diff
@@ -26,10 +26,10 @@
...
-        - [`spec.placement.clusterNames` is not provided, `spec.placement.clusterSelector` is provided but empty](#specplacementclusternames-is-not-provided-specplacementclusterselector-is-provided-but-empty)
-        - [`spec.placement.clusterNames` is not provided, `spec.placement.clusterSelector` is provided and not empty](#specplacementclusternames-is-not-provided-specplacementclusterselector-is-provided-and-not-empty)
+        - [Neither `spec.placement.clusters` nor `spec.placement.clusterSelector` is provided](#neither-specplacementclusters-nor-specplacementclusterselector-is-provided)
+        - [Both `spec.placement.cluster` and `spec.placement.clusterSelector` are provided](#both-specplacementclusters-and-specplacementclusterselector-are-provided)
+        - [`spec.placement.cluster` is not provided, `spec.placement.clusterSelector` is provided but empty](#specplacementclusters-is-not-provided-specplacementclusterselector-is-provided-but-empty)
+        - [`spec.placement.cluster` is not provided, `spec.placement.clusterSelector` is provided and not empty](#specplacementclusters-is-not-provided-specplacementclusterselector-is-provided-and-not-empty)
```
Comment (pmorie) [inline docs/userguide.md]: here, it should be `spec.placement.clusters`

Comment (marun) [reply docs/userguide.md]: oops

Comment (marun) [reply docs/userguide.md]: Done

**Comment by marun** [comment]
Rebased

**Comment by shashidharatd** [comment]
/lgtm
Thanks!

**Pull request merged**